### ISSUE TITLE
Add option to skip unnecessary select after create and return operation status instead

### ISSUE THREAD

**Issue by ZiiMakc** [prisma/prisma#4246]
## Problem

Every time you do operation with prisma-client it will make unnecessary select and return value

Query:
```
db.log
   .create({
      data: {
         smth: 1
      },
      select: { uuid: true }, // unnecessary select, even if you delete this line
   })
```

Logs:
```
prisma:query COMMIT
prisma:query BEGIN
prisma:query INSERT INTO "public"."log" ("smth") VALUES ($1) RETURNING "public"."log"."uuid"
prisma:query SELECT "public"."log"."uuid" FROM "public"."log" WHERE "public"."log"."uuid" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
```

## Suggested solution

Add option to return only operation status - success or error, skipping select after insert or update

```
db.log
   .create({
      data: {
         smth: 1
      },
      select: false
   })
```

**Comment by xiaoyu-tamu** [comment]
I experience the same for upsert

Code
```


      await this.prisma.product.upsert({
        select: { id: true },
        where: { id: product.id },
        update: { id: product.id },
        create: { id: product.id },
      });

```

Logs
```

prisma:query BEGIN
prisma:query SELECT "public"."product"."id" FROM "public"."product" WHERE "public"."product"."id" = $1 OFFSET $2
prisma:query SELECT "public"."product"."id" FROM "public"."product" WHERE "public"."product"."id" = $1
prisma:query UPDATE "public"."product" SET "gtin" = $1, "unit_price" = $2, ...and more WHERE "public"."product"."id" IN ($21)
prisma:query SELECT "public"."product"."id" FROM "public"."product" WHERE "public"."product"."id" = $1 LIMIT $2 OFFSET $3
prisma:query COMMIT
```

**Comment by coryvirok** [comment]
I am also seeing 2 `SELECT`s, 1 `UPDATE`, followed by another `SELECT` the same way that @xiaoyu-tamu reported above.
Are these 4 queries the intended behavior for `upsert()`?

This should be doable with at most 2 queries, (assuming you want to return the rows values after the `UPDATE`):

```js

const id = 1
const col1 = 'foo'
const col2 = 'bar'

// First query (note: using MySQL syntax)
await client.product.$queryRaw`
INSERT INTO product (id, col1, col2) VALUES (${id}, ${col1}, ${col2})
ON DUPLICATE KEY UPDATE 
  col1=VALUES(col1)
  col2=VALUES(col2)
`

// Second query
const product1 = await client.product.findUnique({where: {id: 1}})
```

And if you don't want to return the whole row, just omit the second query.

**Comment by PranavSathy** [comment]
Is there any movement on this? Not sure if this is as relevant, but those select's (after insert even though I haven't specified a `select` argument) do eat into the transaction buffer limit.

**Comment by rinogo** [comment]
Very much in favor of this. I was going to create the issue if it didn't exist.

For what it's worth, I don't mind if the default behavior of `create()` is to return the newly-created row. Until this feature is available, I'll have to resort to (ab)using `$executeRaw`.

**Comment by leoantony72** [comment]
Is there any solution to this , except specifying one field 😐

**Comment by rinogo** [comment]
@leoantony72 I just use `$executeRaw` instead to avoid the extra queries.

**Comment by yousufiqbal** [comment]
Make it optional plz. Not required in my use case at all.

**Comment by rohanrajpal** [comment]
Would appreciate an option to remove the unnecessary select

**Comment by drevisios** [comment]
@matthewmueller 
I would like to add that it is not only an issue of performance and optimisations as in some cases it breaks functionality.

In my case where we use Postgres and Row Level Security (RLS), these "unnecessary" selects cause many problems in the flows and constantly require workarounds to bypass RLS.

As an example:
```sql

-- db admin
create table rls_test( id BIGSERIAL PRIMARY KEY);
create policy rls_test_sel_policy on rls_test for select to app using (false);
create policy rls_test_ins_policy on rls_test for insert to app with check (true);

alter table rls_test ENABLE ROW LEVEL SECURITY;
alter table rls_test FORCE ROW LEVEL SECURITY;

-- as user app
insert into rls_test default values; -- works
insert into rls_test default values returning id; -- fails

```
This is a dummy example (can be solved just with DB Roles) just to demonstrate the issue, as the "returning id" clause violates the select policy.

Also in the case of FKs, the  ``` { connect:{ id: 1 } }``` select also violates security policies that may have been applied. In the connect method I would strongly suggest allowing an "unsafe" or "unchecked" connection.

I can guess removing these selects may break things in the generic way that Prisma works but in our case this is a functionality problem and not a performance one.

**Comment by joshkaplan** [comment]
Additional note: this also applies to `DELETE` (it runs the select before the delete).

Big +1 for this request -- it should be a no-brainer to let users avoid extra queries without having to use an escape-hatch like `$executeRaw`. Every query adds load and latency, and best practice is to minimize both unless it's needed / it's something the user explicitly wants.

**Comment by thilllon** [comment]
+1
Because of this, it shows a significant performance degradation compared to TypeORM. By using `$executeRaw`, I lost the benefits of using Prisma some way.

**Comment by thilllon** [comment]
+1
Because of this, it shows a significant performance degradation compared to TypeORM. By using `$executeRaw`, I lost the benefits of using Prisma in some way.

**Comment by josethz00** [comment]
+1, this is slowing down my queries, how can I just insert new data without having to create more?

**Comment by janpio** [comment]
Can you elaborate exactly the problem this causes for RLS? I am not sure I follow.

**Comment by janpio** [comment]
Yes it does.

**Comment by BMorearty** [comment]
I don’t use RLS but I would also like to not see an extra select on every create, update, or upsert. There’s no need for this extra perf hit and DB load in the common case where the database doesn’t have server-side logic such as triggers that generate field values. 

But if you require me to remember to add `select: false` to every query, I’m going to frequently forget. I’d like to be able to make it the default, either at the whole application level per model. 

E.g., a block attribute could do the trick. Naming is hard. Maybe `@@noSelectAfterWrite` isn’t the right name but this is the idea:

```prisma
model User {
  id    Int @id @default(autoincrement())
  email String

  @@noSelectAfterWrite
}
```

**Comment by bhavesh-chaudhari** [comment]
Is there any update on this issue? I don't want to select anything after update.

**Comment by ajhollowayvrm** [comment]
+1 on this. Currently I've changed my update to only return one field but that's just a workaround. Often the select is the more costly than the create/update itself.

**Comment by hbriese** [comment]
> Can you elaborate exactly the problem this causes for RLS? I am not sure I follow.

Say I want to allow the public to `INSERT` but not `SELECT` a given row meet some condition (e.g. they are a user of the account).
The trailing `RETURNING` requires that the public is permitted to `SELECT` that row as well, requiring the following policy changes:

```diff SQL
CREATE POLICY public_insert ON "Account" FOR INSERT 
    WITH CHECK (true);

+CREATE POLICY public_view ON "Account" FOR SELECT
+    USING (true);

-CREATE POLICY member_view ON "Account" FOR SELECT
-    USING (is_user_account("id"));

CREATE POLICY member_update ON "Account" FOR UPDATE
    USING (is_user_account("id"));
```

**Comment by jonorozcoc** [comment]
Any update on this?

**Comment by stemaDev** [comment]
+1

**Comment by lucashfreitas** [comment]
+1 this is an issue for applications that run Prisma and handled thousand/millions of requests. Seen a lot of people refactoring their inserts/updates otherwise an extra select after every insert/update could cost a lot. Would be good to have this defaulted to not return anything unless the user specifies/intend or maybe have a configuration in the Prisma client generator to enable/disable that.

**Comment by josethz00** [comment]
+1

Relational databases have consistency, its information can almost always be trusted, there's no need to query the database again only to return the same things that were just created

**Comment by azzadrood** [comment]
+1

**Comment by etcimon** [comment]
https://dev.to/josethz00/benchmark-prisma-vs-typeorm-3873

Prisma is over 4,000x slower than typeorm on create workloads because of this

https://res.cloudinary.com/practicaldev/image/fetch/s--zvI0qHD9--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_auto%2Cw_880/https://dev-to-uploads.s3.amazonaws.com/uploads/articles/qfvpaw4xz434xtl2t34b.png

**Comment by PhelixTaken** [comment]
I wish I could skip the `SELECT ...` function when checking if a record exists, without having to retrieve all of its data. Currently, the function selects everything by default, even when I only need to verify its presence. Unfortunately, I can't pass an empty object to the `select` parameter; otherwise it throws an error that I must specify what I want to select.

**Comment by Jun4928** [comment]
+1 Not yet on the roadmap?

**Comment by stemaDev** [comment]
as a workaround we could use createmany, right?

**Comment by jamiter** [comment]
For a current scenario I've also switched to using `updateMany` instead of `update` to prevent the extra select. A wast of resourcing if you're not interested in the result IMO.

A global setting to disable this behaviour would also be nice, instead of a per query solution. IMO it's bad practice to always do an extra select.

**Comment by hbriese** [comment]
> I wish I could skip the `SELECT ...` function when checking if a record exists, without having to retrieve all of its data. Currently, the function selects everything by default, even when I only need to verify its presence. Unfortunately, I can't pass an empty object to the `select` parameter; otherwise it throws an error that I must specify what I want to select.

Use `select: null`

**Comment by reorx** [comment]
@jamiter I've tried `updateMany`, the select still goes on, no even a single one was reduced.

The most ridiculous happens when the model has a self-relation, it will select **3 times** before the `UPDATE` statement is executed, causing random `Expected parent IDs to be set when ordering by parent ID` error with no clue. I've read the code of the underlying Rust query engine, and was sure it's a bug not relate to the database but the query engine iteself. But I won't bother submitting a new issue since there're so many unresolved ones （#9875, #15745, #17316, #17776). It's always the case that the developers asks the issue owner to provide a reproducable exmaple, and then response that they cannot reproduce, leaving the issue untracked for months. Surely the errors are uneasy to reproduce, that doesn't mean the issues should be ignored, but a how obscure and terrible the bug is.

I'm grateful that Prisma saved us a lot of time for rapid development, but I really wish its developers can have more attention on improving the SQL level logics. In my understanding, being an ORM means you have the best practices of all kinds of stuff related to SQL and database, that's the reason for the community to choose you in the long run.

**Comment by Songkeys** [comment]
@reorx Totally agree with you.

(Off-topic ranting below)

There are many unsolved issues with many votes and requests but no developer has been participating in. The only comment you may get is some product manager asking what the real world scenarios you are facing and no response anymore. Based on the current roadmap, they were spending most of time working on improving the performance for serverless and data proxy product but ignoring the basic SQL features and bugs. I understand that they have their own business plan and have to focus on it. And sometimes I really wanted to help but don't know writing rust. I appreciate the brilliant DX of Prisma but just keep getting blocked by issues like this for years. Don't know if there's an alternative ORM to try. :(

**Comment by coryvirok** [comment]
This is likely going to be an unsatisfying recommendation but all those who have felt the pain for long enough will agree - just use a query builder. Knex is great and while there's marginally more boilerplate, it's easy and extensible. 

I use Knex + sql-ts + dmate and I've never looked back (to Prisma.)

**Comment by yeliex** [comment]
@janpio could anyone in dev tema pay attention to this, it is a very basic sql query feature

**Comment by luxaritas** [comment]
Also - the use of `RETURNING` also does not help with mysql

**Comment by alonbilu** [comment]
+1 
This is causing a major performance issue. I need to write a large amounts of data (few hundred rows per second) and this slows down things drastically. In my use case, I have no need for either initial select nor the final 'return'. I just want to do as fast as possible write/update to the db

**Comment by theodiablo** [comment]
Same than many above, we use MySQL so the `RETURNING` feature introduced in Prisma 5.1.0 isn't helpful for us. We'd like to only insert rows and we don't use the result, so we have this unnecessary `select` after a `.create` or `.update` which is definitely hurting our performance when we have a lot of traffic.

I hear that we can use raw SQL to do that, but the whole point we're using Prisma is to avoid writing plain SQL and don't have to update all our queries when doing DB changes.

I guess that we'll have to migrate to a query builder instead, unless there are any plans from the Prisma team to address this? Given that this issue has been opened since 2020, I'm not sure it is quite a high priority yet on the roadmap 🤔
CC @janpio

**Comment by raress96** [comment]
I still can't believe this bug was not fixed. And yes, it is a bug IMO. When you create something, you expect it to be created and that's it. If you want to return fields then you should use `include` and `select`.
This is a HUGE problem in Microservice type application that need to handle a lot of traffic to the database, every additional SELECT hurts performance. RETURNING helps, but it is still unnecessary in a lot of cases.

**Comment by jpenna** [comment]
I think the solution here should be simple: add the option to set `select: null` and then nothing is returned. If `select: undefined`, then things still work the same, returning everything.

**Comment by lzw1926** [comment]
> I think the solution here should be simple: add the option to set `select: null` or `select: false` and then nothing is returned. If `select: undefined`, then things still work the same, returning everything.

+1, and any updates?

**Comment by mykhailo-monchak** [comment]
Do we have an explanation of the rationale behind this approach and the reasons why it cannot be modified?

**Comment by dmytrotree** [comment]
It's 2025 now, any updates?

**Comment by clemens** [comment]
+1.

Ideally, this should be available not only for `create`, but actually all flavors of manipulating methods (`create`, `createMany`, `update`, `updateMany`, `delete`, `deleteMany`, `upsert`).

One could even argue that all of these should default to not returning any data and there should be a pattern like `...AndReturn` (like it exists for `createManyAndReturn` and `updateManyAndReturn`) for the sake of consistency or that these `...AndReturn` methods should be deprecated in favor of always passing something like `select: Prisma.NONE` (for nothing) or `select: { foo: true, bar: true }` (for a selection of fields).

**Comment by coryvirok** [comment]
FWIW, I switched to Drizzle a couple years ago in large part due to this since it shows a lack of performance/security mindedness to not have this implemented after so many years.

**Comment by LCozzoli** [comment]
up

**Comment by danbord** [comment]
+1

**Comment by raress96** [comment]
Is this still not implemented? :sigh

**Comment by roda-jason** [comment]
Would using createMany (with a single item in the data array) solve for this?

**Comment by caioedut** [comment]
> Would using createMany (with a single item in the data array) solve for this?

Only if you don't have nested creates (with relations).

**Comment by franwerner** [comment]
+1

**Comment by vladyslav-n** [comment]
If it gets implemented it would be really awesome! This one's really a performance bottleneck often and it's kinda silly :)

**Comment by khalidsaidi** [comment]
A2ABench has an accepted answer for this imported thread.
- Thread: https://a2abench-api.web.app/q/cmmlg23z700hnuexhplxj38xt
- Accepted at: 2026-03-25T19:13:07.634Z
- Accepted answer agent: `ai village`
- Answer preview: "No matching threads were found for this query."

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319
